File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/AndroidMediaPlayer.java
package com.github.anrimian.musicplayer.data.controllers.music.players;

import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;

import com.github.anrimian.musicplayer.data.controllers.music.equalizer.EqualizerController;
import com.github.anrimian.musicplayer.data.controllers.music.error.PlayerErrorParser;
import com.github.anrimian.musicplayer.data.models.composition.source.UriCompositionSource;
import com.github.anrimian.musicplayer.data.storage.source.CompositionSourceProvider;
import com.github.anrimian.musicplayer.data.utils.rx.RxUtils;
import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics;
import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.composition.source.LibraryCompositionSource;
import com.github.anrimian.musicplayer.domain.models.player.error.ErrorType;
import com.github.anrimian.musicplayer.domain.models.player.events.ErrorEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.FinishedEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PreparedEvent;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;
import io.reactivex.rxjava3.subjects.PublishSubject;

import static android.media.MediaPlayer.MEDIA_ERROR_MALFORMED;
import static android.media.MediaPlayer.MEDIA_ERROR_UNSUPPORTED;

public class AndroidMediaPlayer implements AppMediaPlayer {

    private final BehaviorSubject<Long> trackPositionSubject = BehaviorSubject.create();
    private final PublishSubject<PlayerEvent> playerEventSubject = PublishSubject.create();

    private final Context context;
    private final Scheduler scheduler;
    private final CompositionSourceProvider sourceRepository;
    private final PlayerErrorParser playerErrorParser;
    private final Analytics analytics;
    private final EqualizerController equalizerController;

    private final MediaPlayer mediaPlayer;

    @Nullable
    private Disposable trackPositionDisposable;

    @Nullable
    private Disposable preparationDisposable;

    @Nullable
    private CompositionSource currentComposition;

    private boolean isSourcePrepared = false;
    private boolean playWhenReady = false;
    private boolean isPlaying = false;

    @Nullable
    private ErrorType previousErrorType;

    //problem with error case(file not found), multiple error events
    public AndroidMediaPlayer(Context context,
                              Scheduler scheduler,
                              CompositionSourceProvider sourceRepository,
                              PlayerErrorParser playerErrorParser,
                              Analytics analytics,
                              EqualizerController equalizerController) {
        this.context = context;
        this.scheduler = scheduler;
        this.sourceRepository = sourceRepository;
        this.playerErrorParser = playerErrorParser;
        this.analytics = analytics;
        this.equalizerController = equalizerController;
        mediaPlayer = new MediaPlayer();
        mediaPlayer.setOnCompletionListener(mediaPlayer -> {
            if (currentComposition != null) {
                playerEventSubject.onNext(new FinishedEvent(currentComposition));
            }
        });
        mediaPlayer.setOnErrorListener((mediaPlayer, what, extra) -> {
            sendErrorEvent(what, extra);
            return false;
        });

        try {
            equalizerController.attachEqualizer(mediaPlayer.getAudioSessionId());
        } catch (IllegalStateException ignored) {}
    }

    @Override
    public Observable<PlayerEvent> getEventsObservable() {
        return playerEventSubject;
    }

    @Override
    public void prepareToPlay(CompositionSource composition,
                              long startPosition,
                              @Nullable ErrorType previousErrorType) {
        this.currentComposition = composition;
        this.previousErrorType = previousErrorType;
        RxUtils.dispose(preparationDisposable);
        preparationDisposable = Single.fromCallable(() -> composition)
                .flatMapCompletable(this::prepareMediaSource)
                .doOnEvent(t -> onCompositionPrepared(t, startPosition))
                .onErrorComplete()
                .subscribeOn(scheduler)
                .subscribe();
    }

    @Override
    public void stop() {
        if (!isPlaying) {
            return;
        }
        if (isSourcePrepared) {
            seekTo(0);
        }
        stopTracingTrackPosition();
        if (isSourcePrepared) {
            pausePlayer();
        }
        isPlaying = false;
        playWhenReady = false;
    }

    @Override
    public void resume() {
        if (isPlaying) {
            return;
        }
        if (isSourcePrepared) {
            start();
        }
        playWhenReady = true;
    }

    @Override
    public void pause() {
        if (!isPlaying) {
            return;
        }
        stopTracingTrackPosition();
        pausePlayer();
        isPlaying = false;
        playWhenReady = false;
    }

    @Override
    public void seekTo(long position) {
        try {
            if (isSourcePrepared) {
                mediaPlayer.seekTo((int) position);
            }
        } catch (IllegalStateException ignored) {}
        trackPositionSubject.onNext(position);
    }

    @Override
    public void setVolume(float volume) {
        try {
            mediaPlayer.setVolume(volume, volume);
        } catch (IllegalStateException ignored) {}
    }

    @Override
    public Observable<Long> getTrackPositionObservable() {
        return trackPositionSubject;
    }

    @Override
    public Single<Long> getTrackPosition() {
        return Single.fromCallable(() -> {
            if (currentComposition == null) {
                return 0L;
            }
            try {
                return (long) mediaPlayer.getCurrentPosition();
            } catch (IllegalStateException e) {
                return 0L;
            }
        });
    }

    @Override
    public Single<Long> seekBy(long millis) {
        return getTrackPosition()
                .map(currentPosition -> {
                    try {
                        long targetPosition = currentPosition + millis;
                        if (targetPosition < 0) {
                            targetPosition = 0;
                        }
                        if (targetPosition > mediaPlayer.getDuration()) {
                            return currentPosition;
                        }
                        seekTo(targetPosition);
                        return targetPosition;

                    } catch (IllegalStateException ignored) {
                        return currentPosition;
                    }
                });
    }

    @Override
    public void release() {
        equalizerController.detachEqualizer();
        stopTracingTrackPosition();
        mediaPlayer.release();
    }

    private void startTracingTrackPosition() {
        stopTracingTrackPosition();
        trackPositionDisposable = Observable.interval(0, 1, TimeUnit.SECONDS)
                .observeOn(scheduler)
                .flatMapSingle(o -> getTrackPosition())
                .subscribe(trackPositionSubject::onNext);
    }

    private void stopTracingTrackPosition() {
        if (trackPositionDisposable != null) {
            trackPositionDisposable.dispose();
            trackPositionDisposable = null;
        }
    }

    private void onCompositionPrepared(Throwable throwable, long startPosition) {
        if (throwable == null) {
            seekTo(startPosition);
            playerEventSubject.onNext(new PreparedEvent(currentComposition));
        } else {
            seekTo(0);
            pausePlayer();
            sendErrorEvent(throwable);
        }
    }

    private void sendErrorEvent(int what, int playerError) {
        if (currentComposition != null) {
            playerEventSubject.onNext(new ErrorEvent(
                    getErrorTypeFromPlayerError(what, playerError),
                    currentComposition)
            );
        }
    }

    private ErrorType getErrorTypeFromPlayerError(int what, int playerError) {
        switch (playerError) {
            case MEDIA_ERROR_UNSUPPORTED:
            case MEDIA_ERROR_MALFORMED: {
                return ErrorType.UNSUPPORTED;
            }
            default: {
                analytics.logMessage("unknown player error, what: " + what + ", extra: " + playerError);
                return ErrorType.IGNORED;
            }
        }
    }

    private void sendErrorEvent(Throwable throwable) {
        if (currentComposition != null) {
            ErrorType errorType;
            if (throwable instanceof IOException && previousErrorType == ErrorType.UNSUPPORTED) {
                errorType = ErrorType.UNSUPPORTED;
                previousErrorType = null;
            } else {
                errorType = playerErrorParser.getErrorType(throwable);
            }
            playerEventSubject.onNext(new ErrorEvent(errorType, currentComposition));
        }
    }

    private Completable prepareMediaSource(CompositionSource composition) {
        return prepareMediaSourceInternal(composition)
                .doOnSubscribe(d -> isSourcePrepared = false)
                .doOnComplete(this::onSourcePrepared);
    }

    private Completable prepareMediaSourceInternal(CompositionSource composition) {
        if (composition instanceof LibraryCompositionSource) {
            long id = ((LibraryCompositionSource) composition).getComposition().getId();
            return sourceRepository.getCompositionFileDescriptorSingle(id)
                    .doOnSuccess(fileDescriptor -> {
                        mediaPlayer.reset();
                        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
                        mediaPlayer.setDataSource(fileDescriptor);
                        mediaPlayer.prepare();
                    })
                    .ignoreElement();
        }
        if (composition instanceof UriCompositionSource) {
            return Single.fromCallable(((UriCompositionSource) composition)::getUri)
                    .doOnSuccess(uri -> {
                        mediaPlayer.reset();
                        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
                        mediaPlayer.setDataSource(context, uri);
                        mediaPlayer.prepare();
                    })
                    .ignoreElement();

        }
        throw new IllegalArgumentException("unknown composition source");
    }

    private void onSourcePrepared() {
        if (playWhenReady) {
            start();
        }
        isSourcePrepared = true;
    }

    private void pausePlayer() {
        pause(mediaPlayer);
    }

    private void start() {
        start(mediaPlayer);
        startTracingTrackPosition();
        isPlaying = true;
    }

    private void pause(MediaPlayer mediaPlayer) {
        try {
            if (mediaPlayer.isPlaying()) {
                mediaPlayer.pause();
            }
        } catch (Exception ignored) {}
    }

    private void start(MediaPlayer mediaPlayer) {
        try {
            mediaPlayer.start();
        } catch (IllegalStateException ignored) {}
    }

}

File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/AndroidMediaPlayer.kt
package com.github.anrimian.musicplayer.data.controllers.music.players

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer
import android.os.Build
import com.github.anrimian.musicplayer.data.controllers.music.equalizer.EqualizerController
import com.github.anrimian.musicplayer.data.controllers.music.players.utils.MediaPlayerDataSourceBuilder
import com.github.anrimian.musicplayer.data.models.composition.source.UriContentSource
import com.github.anrimian.musicplayer.data.utils.hasPersistedReadPermission
import com.github.anrimian.musicplayer.domain.interactors.analytics.Analytics
import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource
import com.github.anrimian.musicplayer.domain.models.composition.content.NoReadPermissionException
import com.github.anrimian.musicplayer.domain.models.composition.content.RelaunchSourceException
import com.github.anrimian.musicplayer.domain.models.composition.content.UnknownPlayerException
import com.github.anrimian.musicplayer.domain.models.composition.content.UnsupportedSourceException
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance
import com.github.anrimian.musicplayer.domain.models.player.events.MediaPlayerEvent
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.subjects.PublishSubject
import java.io.IOException
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class AndroidMediaPlayer(
    private val context: Context,
    private val ioScheduler: Scheduler,
    private val equalizerController: EqualizerController,
    private val sourceBuilder: MediaPlayerDataSourceBuilder,
    private val analytics: Analytics
) : AppMediaPlayer {

    private val playerEventsSubject = PublishSubject.create<MediaPlayerEvent>()

    private val mediaPlayer = MediaPlayer().apply {
        setOnCompletionListener { onPlaybackFinished() }
        setOnErrorListener { _, what, extra ->
            val ex = createExceptionFromPlayerError(what, extra)
            playerEventsSubject.onNext(MediaPlayerEvent.Error(ex))
            false
        }
    }

    private var currentSource: CompositionContentSource? = null

    private var isSourcePrepared = false
    private var isPlaying = false

    private var previousException: Exception? = null
    private var postponedPosition: Long? = null

    private var volume = 1f
    private var leftVolume = 1f
    private var rightVolume = 1f

    private var trackEndMonitorDisposable: Disposable? = null
    private val isFinished = AtomicBoolean(false)

    override fun prepareToPlay(
        source: CompositionContentSource,
        previousException: Exception?,
    ): Completable {
        stopTrackEndMonitor()
        isFinished.set(false)
        currentSource = source
        postponedPosition = null
        this.previousException = previousException
        return prepareMediaSource(source)
            .doOnSubscribe { isSourcePrepared = false }
            .doOnComplete {
                isSourcePrepared = true
                postponedPosition?.let(this::seekTo)
            }
            .onErrorResumeNext { t -> Completable.error(mapPrepareException(t, source)) }
            .subscribeOn(ioScheduler)
    }

    override fun stop() {
        if (!isPlaying) {
            return
        }
        stopTrackEndMonitor()
        if (isSourcePrepared) {
            seekTo(0)
        }
        if (isSourcePrepared) {
            pausePlayer()
        }
        isPlaying = false
    }

    override fun resume() {
        if (isPlaying) {
            return
        }
        if (isSourcePrepared) {
            start()
        }
    }

    override fun pause() {
        if (!isPlaying) {
            return
        }
        stopTrackEndMonitor()
        pausePlayer()
        isPlaying = false
    }

    override fun seekTo(position: Long) {
        try {
            if (isSourcePrepared) {
                synchronized(mediaPlayer) {
                    mediaPlayer.seekTo(position.toInt())
                    // If logical state is playing but physical playback stopped (e.g. track ended),
                    // we must restart it to handle track repeats correctly
                    if (isPlaying && !mediaPlayer.isPlaying) {
                        mediaPlayer.start()
                        isFinished.set(false)
                        startTrackEndMonitor()
                    }
                }
            } else if (currentSource != null) {
                postponedPosition = position
            }
        } catch (ex: Exception) {
            val exception = if (ex is SecurityException) {
                ex
            } else {
                RelaunchSourceException(ex)
            }
            playerEventsSubject.onNext(MediaPlayerEvent.Error(exception))
        }
    }

    override fun setVolume(volume: Float) {
        this.volume = volume
        applyVolume()
    }

    override fun getTrackPositionObservable(): Observable<Long> {
        return Observable.interval(0, 50, TimeUnit.MILLISECONDS)
            .observeOn(ioScheduler)
            .flatMapSingle { getTrackPosition() }
    }

    override fun getTrackPosition(): Single<Long> {
        return Single.fromCallable {
            if (!isSourcePrepared) {
                return@fromCallable 0L
            }
            try {
                synchronized(mediaPlayer) {
                    return@fromCallable mediaPlayer.currentPosition.toLong()
                }
            } catch (_: IllegalStateException) {
                return@fromCallable 0L
            }
        }
    }

    override fun getDuration(): Single<Long> {
        return Single.fromCallable {
            if (!isSourcePrepared) {
                return@fromCallable 0L
            }
            try {
                synchronized(mediaPlayer) {
                    return@fromCallable mediaPlayer.duration.toLong()
                }
            } catch (_: IllegalStateException) {
                return@fromCallable 0L
            }
        }
    }

    override fun setPlaybackSpeed(speed: Float) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            try {
                synchronized(mediaPlayer) {
                    val isPlaying = mediaPlayer.isPlaying
                    mediaPlayer.playbackParams = mediaPlayer.playbackParams.setSpeed(speed)
                    if (!isPlaying) {
                        mediaPlayer.pause()
                    }
                }
            } catch (_: IllegalStateException) {
            } catch (_: IllegalArgumentException) {
            } //IllegalArgumentException - reset ui
        }
    }

    override fun setSkipSilenceEnabled(enabled: Boolean) {
        //not supported
    }

    override fun release() {
        synchronized(mediaPlayer) {
            stopTrackEndMonitor()
            isSourcePrepared = false
            currentSource = null
            equalizerController.detachEqualizer()
            mediaPlayer.release()
        }
    }

    override fun getPlayerEventsObservable(): Observable<MediaPlayerEvent> {
        return playerEventsSubject
    }

    override fun getSpeedChangeAvailableObservable(): Observable<Boolean> {
        return Observable.fromCallable { Build.VERSION.SDK_INT >= Build.VERSION_CODES.M }
    }

    override fun setSoundBalance(soundBalance: SoundBalance) {
        leftVolume = soundBalance.left
        rightVolume = soundBalance.right
        applyVolume()
    }

    private fun applyVolume() {
        val leftOutput = volume * leftVolume
        val rightOutput = volume * rightVolume
        try {
            synchronized(mediaPlayer) {
                mediaPlayer.setVolume(leftOutput, rightOutput)
            }
        } catch (_: IllegalStateException) {}
    }

    private fun createExceptionFromPlayerError(what: Int, playerError: Int): Exception {
        return when (playerError) {
            MediaPlayer.MEDIA_ERROR_UNSUPPORTED,
            MediaPlayer.MEDIA_ERROR_MALFORMED,
            -> UnsupportedSourceException()
            else -> {
                val message = "unknown media player error, what: $what, extra: $playerError"
                UnknownPlayerException(message)
            }
        }
    }

    private fun mapPrepareException(throwable: Throwable, source: CompositionContentSource): Throwable {
        if (throwable is IOException) {
            if (source is UriContentSource && !source.uri.hasPersistedReadPermission(context)) {
                return NoReadPermissionException(throwable)
            }
        }
        
        if (previousException is UnsupportedSourceException) {
            previousException = null
            return UnsupportedSourceException()
        }

        return throwable
    }

    private fun prepareMediaSource(source: CompositionContentSource): Completable {
        return Completable.create { emitter ->
            synchronized(mediaPlayer) {
                try {
                    mediaPlayer.reset()
                    mediaPlayer.setAudioAttributes(
                        AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .setUsage(AudioAttributes.USAGE_MEDIA)
                            .build()
                    )
                    sourceBuilder.setMediaSource(mediaPlayer, source)

                    mediaPlayer.setOnPreparedListener {
                        if (!emitter.isDisposed) {
                            emitter.onComplete()
                        }
                    }
                    mediaPlayer.setOnErrorListener { _, what, extra ->
                        isSourcePrepared = false
                        val ex = createExceptionFromPlayerError(what, extra)
                        if (!emitter.isDisposed) {
                            emitter.tryOnError(ex)
                        }
                        true // return true to ignore error
                    }
                    mediaPlayer.prepareAsync()
                } catch (t: Throwable) {
                    if (!emitter.isDisposed) {
                        emitter.tryOnError(t)
                    }
                }
            }
        }
    }

    private fun pausePlayer() {
        try {
            synchronized(mediaPlayer) {
                if (mediaPlayer.isPlaying) {
                    mediaPlayer.pause()
                    equalizerController.detachEqualizer()
                }
            }
        } catch (_: Exception) {}
    }

    private fun start() {
        synchronized(mediaPlayer) {
            try {
                mediaPlayer.start()
                startTrackEndMonitor()
                equalizerController.attachEqualizer(mediaPlayer.audioSessionId)
                isPlaying = true
            } catch (_: IllegalStateException) {}
        }
    }

    private fun onPlaybackFinished() {
        if (isFinished.compareAndSet(false, true)) {
            stopTrackEndMonitor()
            playerEventsSubject.onNext(MediaPlayerEvent.Finished)
        }
    }

    private fun startTrackEndMonitor() {
        stopTrackEndMonitor()
        trackEndMonitorDisposable = getDuration()
            .flatMapObservable { duration ->
                if (duration <= 0) {
                    return@flatMapObservable Observable.empty()
                }
                Observable.interval(0, 500, TimeUnit.MILLISECONDS, ioScheduler)
                    .flatMapSingle { getTrackPosition() }
                    .doOnNext { position ->
                        if (position >= duration - PLAYBACK_END_TOLERANCE_WINDOW_MS) {
                            onPlaybackFinished()
                        }
                    }
            }
            .doOnError { t -> analytics.processNonFatalError(t, "Track end monitor") }
            .onErrorComplete()
            .subscribe()
    }

    private fun stopTrackEndMonitor() {
        trackEndMonitorDisposable?.dispose()
        trackEndMonitorDisposable = null
    }


    private companion object {
        /**
         * A tolerance window in milliseconds to account for inaccuracies in MediaPlayer's
         * position reporting near the end of a track.
         */
        private const val PLAYBACK_END_TOLERANCE_WINDOW_MS = 200L
    }

}
